### ISSUE TITLE
New Features

### ISSUE THREAD

**Issue by emanelfy** [cds-snc/ircc-rescheduler#398]
Adding two new features to the service:
1. "Reschedule my family"
2. "These dates don't work for me"
3. New "Explanation" page 
4. Review page update

--

### Form Page
Order of elements on the page:

- Name
- PFN
- **Schedule family** 
_Should look like this:_
Field should be the same width as the text field used for the reason at the bottom of the page, but 2/3s of the height. The text field only becomes activated and mandatory if the box is checked (otherwise the field is in a disabled grey state).
<img width="654" alt="screen shot 2018-08-27 at 6 27 18 am" src="https://user-images.githubusercontent.com/35348560/44655192-45869e80-a9c2-11e8-8d3c-ebfe2ad2cb01.png">

cc @zakkain - to supply content + error message for if box is checked, but no input in field.

--
- Email
- Radio buttons
- Details field

### Calendar Page
Should be present on page load (and stay regardless of whether calendar selections are made or not).
Left-aligned with the date-box, and aligned with the dates in the last row of the calendar.

**Mobile:** Stacked above the calendar, below the calendar instruction.

Same size and colour as the "cancel" link at the bottom of the screen. 
<img width="933" alt="screen shot 2018-08-27 at 6 20 24 am" src="https://user-images.githubusercontent.com/35348560/44654920-626ea200-a9c1-11e8-9d74-e672c6816051.png">

### Explanation Page
Should only contain a field box that looks like this:
<img width="645" alt="screen shot 2018-08-28 at 10 44 45 am" src="https://user-images.githubusercontent.com/35348560/44730639-68db4780-aaaf-11e8-96f3-2f072e8e2866.png">

@zakkain to supply error message for if a user tries to continue without entering text in the field, and a warning box message to clarify that their request may be reviewed and that they should attend their original appointment, and when staff will respond to users by.

This also changes the nature of the warning messages from that point forward in the flow (review page, etc).

### Review page update
When moving forward in the flow, the dates that would normally appear under the availability section should be replaced by the text inputted in the field on the page prior.
<img width="900" alt="screen shot 2018-08-28 at 10 47 39 am" src="https://user-images.githubusercontent.com/35348560/44730834-d5564680-aaaf-11e8-8fe0-03f5ba8cf88b.png">
Flow progresses normally from here.

**Comment by emanelfy** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Dates don't work for you? option](https://trello.com/c/mjSMMSBb/317-dates-dont-work-for-you-option)

**Comment by zakkain** [comment]
### Family feature
Checkbox 👉 `I need to reschedule my family too`
Box text  👉 `Provide the name of each family member you want to reschedule.`
Empty box error  👉`You left this blank. Do you want to reschedule any members of your family? Please provide their names.`

### Calendar page
Link on calendar page 👉 `I‘m not available for any these days`

### Explanation page
Page title 👉 `Delay your appointment`
Page description text (cc @emanelfy Hillary and i think we need this disclaimer element.) 👉 `If you aren't available on any of the upcoming days, please provide a reason for the delay. We will reply to you in 1 week to confirm your delay, or ask for more information. There is a chance you will not be granted a delay.`
Box label 👉 `Describe why you need to delay your appointment`
Box description 👉 `Provide enough detail so that staff can understand your situation.`
Empty box error 👉 `Please tell us why you aren’t available on any of the upcoming days.`

### Review page
Warning box (for the new use case, only) 👉 `You should plan to attend your existing appointment until we contact you. This can take up to a week.`

note: the "1 week" language here is pending, Hillary is going to run it by the offices

**Comment by zakkain** [comment]
great feedback, thanks! Updating original post 👍

**Comment by emanelfy** [comment]
@hillaryl & I have just met about the unavailability flow redesign.

**Here's the updated content for the explanation page:** 

**Page title** 👉 Apply for an appointment extension

 **Page description text** 👉 If you cannot attend an appointment on any of the available days, staff may not be able to accommodate your request to reschedule. _[Line about policy, checking with an IRCC local office — Our policy is to delay appointments a maximum of two months, unless ____ reason.]_
We’ll review your request and determine if we can accommodate your unavailability and get back to you within 1 week. 

 **Box label** 👉 When are you unavailable?

 **Box description** 👉 If you haven’t already, please provide a reason for your unavailability. 

 **Empty box error** 👉 Please provide us with more information. 

--
cc @zakkain @pcraig3 @nmakuch @anikbrazeau

**Comment by emanelfy** [comment]
**_Here's the latest round of design updates to the service._**

### Form page update:
On the family text-field, add grey italic hint text inside field saying _“full name, full name, full name”_

### Calendar page updates:
If a user checks the "family option" on the form page, the content on the calendar page should be updated to say
<img width="967" alt="screen shot 2018-09-05 at 5 06 17 pm" src="https://user-images.githubusercontent.com/35348560/45123488-134d0d80-b135-11e8-8950-b57950f9a5e9.png">

Remove "these dates don't work for me" link and insert a check box instead:
<img width="881" alt="screen shot 2018-09-05 at 6 05 10 pm" src="https://user-images.githubusercontent.com/35348560/45123855-43e17700-b136-11e8-89e7-90dc4d2c4156.png">

When the box is checked, the calendar should change to grey and become inactive.
The green button at the bottom of the page should change to say **next** and take users to the explanation page.

**Other interaction design elements on the calendar page to be added:**

- Arrow on the calendar will be green (colour to match our buttons) and jiggle (only on the first page of the calendar)

- the white available date areas on the calendar should be outlined in green (same colour as above)

- the highlight colour on the calendar is now a lighter shade of green instead of grey

- the icon that appears on the non-selectable dates should be red (to match our error messaging)

### Explanation page update:
Include "< Go back" link at the top of the page.
[see previous comment above for copy edits]

### Confirmation page
Copy edit:
**What happens next?**
IRCC will send you _an email with your_ new appointment.

--
cc @hillaryl @zakkain @pcraig3 @nmakuch @anikbrazeau

**Comment by timarney** [comment]
@emanelfy what should the calendar in the disabled state look like?

 "the calendar should change to grey and become inactive."

Something like this?
<img width="481" alt="screen shot 2018-09-10 at 2 27 44 pm" src="https://user-images.githubusercontent.com/62242/45316690-c044d300-b505-11e8-8bd6-4c72ea244ff1.png">

**Comment by pcraig3** [comment]
Hey all,

The copy updates are getting hard to keep track of in this big issue.

Currently we have a bunch of new content going in so I'm going to try and do a master list here.

It's possible to edit this comment directly, so please do that if you want to change some of the text / add new text somewhere.

## Family option

| where is it? | en | fr |
|--------------|----|----|
|  register page: checkbox label   |  I need to reschedule my family too |  J’ai aussi besoin de reporter le rendez-vous de membres de ma famille.  |
|  register page: textarea label   | Provide the full name of each family member you want to reschedule  |  Veuillez indiquer le nom de chaque membre de votre famille pour qui vous désirez reporter le rendez-vous.  |
|  register page: textarea placeholder text  | “Example: Full Name, Full Name, Full Name”  |  « Par exemple : Nom Complet, Nom Complet »  |
|  register page: error description at top of page when checkbox is clicked but no family members are provided  |  Provide the names of your family members  |  Veuillez indiquer les noms des membres de votre famille.  |
|  register page: error label for textarea when checkbox is clicked but no family members are provided  |  You left this blank. Do you want to reschedule any members of your family? Please provide their full names.  |  Vous avez laissé cette case vide. Aimeriez-vous reporter le rendez-vous pour des membres de votre famille? Veuillez indiquer leurs noms.  |
|  register page: error label for textarea if too many words  | There is a limit of 150 words for your family’s names. Please shorten your explanation. |  Désolé, il y a une limite de 150 mots. Veuillez racourcir votre explication.  |
|  register page: error description at top of page when checkbox is not clicked but family members are entered  |  Please confirm you are rescheduling family members.  |  Veuillez confirmer qui vous désirez aussi reporter le rendez-vous pour des membres de votre famille.  |
|  register page: error label for checkbox when checkbox is not clicked but family members are entered  | You must click ‘I need to reschedule my family too’ if you are rescheduling family members. |  Vous devez sélectionner la case si vous désirez aussi reporter le rendez-vous pour des membres de votre famille.  |
|  calendar page: page title  |  Select 3 days you and your family are available between ... |       |
|  review page: row name if family members submitted  |  Family members  |  Les membres de ma famille |


## These dates don't work for me


| where is it? | en | fr |
|--------------|----|----|
|  calendar page: checkbox label   |  I'm not available for any of these days |  Je ne suis pas disponible pour l’ensemble de ces jours  |
|  explanation page: page title   |  Apply for an appointment extension |  Demander une extension de rendez-vous  |
|   explanation page: first sentence   |  If you cannot attend an appointment on any of the available days, staff may not be able to accommodate your request to reschedule. | Si vous ne pouvez pas assister à un rendez-vous les jours disponibles, le personnel peut ne pas être en mesure de répondre à votre demande de replanification.  |
|   explanation page: second sentence  |  Our policy is to delay appointments a maximum of two months. We’ll review your request and determine if we can accommodate your unavailability and get back to you within 1 week. | Notre politique est de reporter les rendez-vous au maximum deux mois. Nous examinerons votre demande et déterminerons si nous pouvons répondre à votre indisponibilité et vous contacter dans un délai d’une semaine.  |
|   explanation page: textarea label |  When are you unavailable?  | Quand êtes-vous indisponible? |
|   explanation page: textarea subheading  |  If you haven’t already, please provide a reason for your unavailability | Si vous ne l'avez pas encore fait, veuillez fournir une raison pour votre indisponibilité |
|   explanation page: textarea empty error  | Please provide us with more information.  | Veuillez nous fournir plus d'informations. |
|   explanation page: textarea too many words error  | Sorry, there‘s a limit of 150 words for this explanation. Please shorten your explanation.  | Désolé, il y a une limite de 150 mots. Veuillez racourcir votre explication. |
|   review page: warning/info message  | You should plan to attend your existing appointment until we contact you. This may take 1 week.  | Vous devriez prévoir assister à votre rendez-vous actuel jusqu’à ce que nous communiquions avec vous. Cela pourrait prendre jusqu’à une semaine. |

**Comment by timarney** [comment]
**Update on this issue:**

- The family option is in (most pieces are live)
- "You and your family" is merged but not live yet.
- 👉 These dates don't work for me see this comment https://github.com/cds-snc/ircc-rescheduler/pull/438#issuecomment-421005773 (also covers the Explanation, Review Page, + updates to the Email templates)
- Animated arrow is live (we can adjust further as need be)

**Still on the todo list:**
- the highlight colour on the calendar is now a lighter shade of green instead of grey
- the icon that appears on the non-selectable dates should be red (to match our error messaging)

**Next Steps:**
This issues covers quite a few things so looking to make sure we have them all covered.

Please review and flag anything that's missing (translations for PR #438 will be added at the end) or that needs further tweaking.

**Comment by zakkain** [comment]
The only bit of extra feedback I have here is the 150-word limit for the explanation. Is that an arbitrary one? The error messaging is fine, and I don't expect many will hit it but... do we need a limit? Can it be something that will catch spam and bot stuff (like, huge strings with no spaces, or like, 10,000 characters...) instead of something a reasonable person might actually hit? Some of the explanation emails IRCC gets are definitely longer than 150 words

**Comment by timarney** [comment]
We can up the limit for that field 

We have the same limit for the family option
https://github.com/cds-snc/ircc-rescheduler/blob/master/web/src/validation.js#L58

The actual # is a character count

```
const TEXTAREA_MAX_CHARS = 1500
```

Let me know what you think a decent character + translated into word count would be for that field.

**Comment by zakkain** [comment]
I looked back at the character-counts of the emails IRCC was getting and they bounce between ~80 words on the low end, to ~350 words on the high end. So, in characters that's around ~2000 on the high end. I'd feel comfortable upping the character count to 2000 — this isn't an open email, it's a targeted field, so that seems appropriate to me.

**Comment by zakkain** [comment]
https://github.com/cds-snc/ircc-rescheduler/issues/452#issuecomment-421466209

**Comment by anikbrazeau** [comment]
where is it? | en | fr
-- | -- | --
register page: checkbox label | I need to reschedule my family too | **J’ai aussi besoin de reporter le rendez-vous pour des membres de ma famille.**
register page: textarea label | Provide the full name of each family member you want to reschedule | **Veuillez indiquer le nom complet de chaque membre de votre famille pour qui vous désirez reporter le rendez-vous.**
register page: textarea placeholder text | For example: “Full Name, Full Name, Full Name” | **par exemple : « Nom Complet, Nom Complet, Nom Complet »**
register page: error description at top of page when checkbox is clicked but no family members are provided | Reschedule family members | Reporter le rendez-vous pour des membres de votre famille
register page: error label for textarea when checkbox is clicked but no family members are provided | You left this blank. Do you want to reschedule any members of your family? Please provide their full names. | **Vous avez laissé cette case vide. Aimeriez-vous reporter le rendez-vous pour des membres de votre famille? Veuillez indiquer leurs noms complets.**
register page: error label for textarea if too many words | There is a limit of 150 words for your family’s names. Please shorten your explanation. | Il y a une limite de 150 mots pour les noms des membres de votre famille. Veuillez raccourcir votre explication.
register page: error description at top of page when checkbox is not clicked but family members are entered | Reschedule family members | Reporter le rendez-vous pour des membres de votre famille
register page: error label for checkbox when checkbox is not clicked but family members are entered | You must click ‘I need to reschedule my family too’ if you are rescheduling family members. | Vous devez sélectionner la case si vous désirez aussi reporter le rendez-vous pour des membres de votre famille.
calendar page: page title | Select 3 days you and your family are available between ... | **Sélectionnez 3 jours entre le ________ et le ________ où vous et votre famille serez disponibles.**
review page: row name if family members submitted | Family members | Les membres de votre famille

**Comment by anikbrazeau** [comment]
where is it? | en | fr
-- | -- | --
calendar page: checkbox label | I'm not available for any of these days | Je ne suis pas disponible pour l’ensemble de ces jours
explanation page: page title | Apply for an appointment extension | **Demander une prolongation du rendez-vous**
explanation page: first sentence | If you can‘t attend an appointment on any of the available days, staff may not be able to accommodate your request. | **Si vous ne pouvez pas assister à un rendez-vous pour l'ensemble des jours disponibles, le personnel pourrait ne pas être en mesure de satisfaire votre demande.**
explanation page: second sentence | Our policy is to delay appointments up to two months. We’ll review your request and determine if we can accommodate your unavailability and get back to you within 1 week. | **Notre politique est de repousser les rendez-vous au maximum deux mois. Nous examinerons votre demande et déterminerons si nous pouvons répondre à votre indisponibilité et vous contacter dans un délai d’une semaine.**
explanation page: textarea label | When are you unavailable? | Quand êtes-vous indisponible?
explanation page: textarea subheading | If you haven’t already, please provide a reason for your unavailability | Si vous ne l'avez pas encore fait, veuillez fournir une raison pour votre indisponibilité
explanation page: textarea empty error | Please provide us with more information. | Veuillez nous fournir plus d'informations.
explanation page: textarea too many words error | Sorry, there‘s a limit of 150 words for this explanation. Please shorten your explanation. | Désolé, il y a une limite de 150 mots. Veuillez racourcir votre explication.
review page: warning/info message | You should plan to attend your existing appointment until we contact you. This may take 1 week. | Vous devriez prévoir assister à votre rendez-vous actuel jusqu’à ce que nous communiquions avec vous. Cela pourrait prendre jusqu’à une semaine.

**Comment by cgovias** [comment]
Hey folks, chiming in here on a couple of design bits. @pcraig3 @emanelfy @timarney @zakkain 

![image](https://user-images.githubusercontent.com/505697/45839962-c005be00-bce3-11e8-9886-d1a90869586b.png)

There are a few problems with the greyed-out form field:

1) hint text – _For example: "Full Name, Full Name, Full Name"_  – shouldn't ever be inside the input field itself. It should sit *above*, between the input field and the input field label.

2) The hint text itself feels problematic in its use of the word "for example". Suggest it be something more like "Please add each family member's full name, separated by a comma." @zakkain over to you on this.

3) we should *not* show the input field itself  *unless* the user selects the tick box. If a user skims the page, the grey input field (another change to the visual language of the interface) stands out way too much. My hunch (research required!) is that users will fixate on it. If the majority of users don't need this, don't show all the bits and pieces related to it. Instead, we should hide the input field entirely. Once the user selects "I need to reschedule my family", then show the input field with appropriate hint text. _Progressive disclosure for the win._

**Comment by emanelfy** [comment]
We actually put this new feature (specifically this hint text) to the test yesterday in our final usability testing session, and saw no issues with the hint text being in the field.

We realize that best practice suggests moving it outside of the box so that it is always present for users, however, we are simply using this to provide clarity and nudge users. 

The field is not validated by how users enter names into it. Whatever they input into the field is fine, will be accepted, and will be displayed to them exactly as they have written it when they review it later/receive a confirmation email. If they decide to enter names with a line break between each, this is also acceptable. We only provided an example so that users who may inquire about how to format/write the names in the field have a bit of guidance.

Furthermore, the hint text remains present even when the field is active/clicked-on, and only disappears when a user actually starts to type. 

Since our users are predominantly English as second-language, we want to keep the example a more visual one and less verbose.

We saw no issues yesterday during testing — however, **style recommendation is to remove the quotation marks around the names (cc @pcraig3),** as one user thought they had to put quotation marks around the names they were including.

**Comment by cgovias** [comment]
@emanelfy Good stuff. Point 3 still stands though: we shouldn't show this input field **unless** the user ticks the "I need to reschedule my family" box.

**Comment by emanelfy** [comment]
@cgovias unfortunately in previous rounds of testing, progressive disclosure did not perform well. The changes on the page are too subtle that users completely miss it and trigger errors. We’d previously discussed this as a team and decided to stick with pre-loading everything on the page and experiment with “disabled-grey” elements instead. So far we haven’t encountered any issues.

Would be wary of rolling out a change of this nature (given that implications could be massive) without having the opportunity to test it with users first.

**Comment by emanelfy** [comment]
@cgovias @pcraig3 — completely agree on the calendar chevron! Green background is definitely more obvious/better treatment overall😃

**Comment by cgovias** [comment]
@emanelfy Yeah, I hear that re: the progressive disclosure. It's the unfortunate reality that when testing these things, every litttle thing matters – position, size, layout, hierarchy, colour, visual, etc. Progressive disclosure is actually the best way forward, but it has to be set up for success – whatever leads to the disclosure needs to be big, clear, obvious, etc. Is there any way to see what the prog. disclosure version looked like?

**Comment by pcraig3** [comment]
To comment on the progressive disclosure point specifically in the context of this product:

- It would mean more dev effort especially around the edge cases (ie,it wouldn't work with our noJS flow.)
- Personally, didn't seen anyone stumble over it in the usability research sessions I attended
- This way works well for assistive tech, since `disabled` fields are skipped

Outside of this project, on the Marketplace we had a flow where additional questions would show up based on your response to prior questions. [Here's an issue I raised](https://github.com/alphagov/govuk_frontend_toolkit/pull/383), and [here're some working examples](http://alphagov.github.io/digitalmarketplace-frontend-toolkit/forms/combinations.html).

General observations:
- People largely didn't like it (granted, we had a *lot* of them)
- The noJS alternative was bad 👇 

![image](https://user-images.githubusercontent.com/2454380/45904953-a7b5a200-bdbc-11e8-9d18-5cc22b048a6e.png)

Not an anti-progressive-disclosure partisan, but, as @emanelfy said, we haven't really seen a problem thus far and personally I wouldn't be confident in changing it shortly before the handoff (especially given the marketplace stuff where a lot of users mentioned they were surprised by all the new textareas showing up).
